index.html

```diff
@@ -204,55 +205,2367 @@ <h3>
+              {{DOMException}}.
+              </li>
+            </ol>
+          </li>
+          <li>Repeat for each |pendingPromise:Promise| of
+          |pendingSendFeatureReportPromises|:
+            <ol>
+              <li>[=Reject=] |pendingPromise| with an "{{AbortError}}"
+              {{DOMException}}.
+              </li>
+            </ol>
+          </li>
+          <li>Repeat for each |pendingPromise:Promise| of
+          |pendingReceiveFeatureReportPromises|:
+            <ol>
+              <li>[=Reject=] |pendingPromise| with an "{{AbortError}}"
+              {{DOMException}}.
+              </li>
+            </ol>
+          </li>
+          <li>Perform the following steps [=in parallel=]:
+            <ol>
+              <li>Invoke the operating system to close the HID device and
+              release any associated resources.
+              </li>
+              <li>Set [=this=].{{HIDDevice/[[state]]}} to `"closed"`.
+              </li>
+              <li>[=Queue a global task=] on the [=relevant global object=] of
+              [=this=] using the [=HID device task source=] to [=resolve=]
+              |promise| with `undefined`.
+              </li>
+            </ol>
+          </li>
+          <li>Return |promise|.
+          </li>
+        </ol>
+      </section>
+      <section>
+        <h3>
+          <dfn>sendReport()</dfn> method
+        </h3>
+        <div class="note">
+          <p>
+            The {{HIDDevice/sendReport()}} method is used to send an output
+            report with the specified |reportId:octet| and |data:BufferSource|.
+            If the [=HID interface=] does not [=use report IDs=], pass 0
+            instead of the report ID.
+          </p>
+          <p>
+            For [=HID interfaces=] that [=use report IDs=], the report ID value
+            0 is reserved and should not be used.
+          </p>
+        </div>
+        <p>
+          The {{HIDDevice/sendReport()}} method steps are:
+        </p>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>If [=this=].{{HIDDevice/[[state]]}} is not `"opened"`, then
+          [=reject=] |promise| with a "{{InvalidStateError}}" {{DOMException}}
+          and return |promise|.
+          </li>
+          <li>If |reportId:octet| is 0 and the [=HID interface=] represented by
+          [=this=] [=uses report IDs=], or if |reportId| is not 0 and the
+          interface does not [=use report IDs=], then [=reject=] |promise| with
+          a "{{TypeError}}" {{DOMException}} and return |promise|.
+          </li>
+          <li>Append |promise| to
+          [=this=].{{HIDDevice/[[pendingSendReportPromises]]}}.
+          </li>
+          <li>Perform the following steps [=in parallel=]:
+            <ol>
+              <li>If |promise| is not in
+              [=this=].{{HIDDevice/[[pendingSendReportPromises]]}}, then abort
+              these steps.
```

Comment:
I think it is fine not to have this logic in here because promises can't be resolved more than once and so these steps become a no-op if they continue running after the device is closed. Importantly, you would need to do this check again after invoking the operating system because that's the truly asynchronous part of this operation.

Another way you could model this is using a list of AbortControllers instead of Promises. That comes with a way of having a set of parallel steps which can react to an abort. It'll be a little weird not to actually expose the AbortController to script but Chromium doesn't actually support aborting these operations so it wouldn't be a good idea to give developers the false sense that that were possible.

Reply:
Removed